### ISSUE TITLE
BUG: signal.resample: inconsistency across dtypes

### ISSUE THREAD

**Issue by mreineck** [scipy/scipy#14569]
When calling `scipy.signal.resample` with the same real-valued input, but provided as arrays of real and complex type, the returned result is different when the length of the resampled array is 2.

#### Reproducing code example:
```
import numpy as np


def test(L,L2):
    import scipy.signal
    a = np.random.random(L)
    k = np.random.random(L)
    x=scipy.signal.resample(a,L2,axis=0,domain='time')
    x2=scipy.signal.resample(a.astype(np.complex128),L2,axis=0,domain='time')
    if np.max(np.abs(x-x2)) > 1e-4:
        print("resampling inconsistency for", L, "->", L2)


for l1 in [2,3,4,5,6,10,11]:
    for l2 in [2,3,4,5,6,10,11]:
        test(l1, l2)
```

#### Output:

```
resampling inconsistency for 3 -> 2
resampling inconsistency for 4 -> 2
resampling inconsistency for 5 -> 2
resampling inconsistency for 6 -> 2
resampling inconsistency for 10 -> 2
resampling inconsistency for 11 -> 2
```

#### Scipy/Numpy/Python version information:

1.6.0 1.19.5 sys.version_info(major=3, minor=9, micro=2, releaselevel='final', serial=0)

**Comment by mreineck** [comment]
As far as I can tell, the real-valued resampling is doing the right thing and the complex-valued one is wrong in the problematic cases.

**Comment by czgdp1807** [comment]
I am able to reproduce the output using the code in the description. Can I investigate further into this issue?

**Comment by czgdp1807** [comment]
An easy fix is to consider only real parts if imaginary parts are all zero. See the diff below.

```diff
diff --git a/scipy/signal/signaltools.py b/scipy/signal/signaltools.py
index bb43c0d25..eacc57d5d 100644
--- a/scipy/signal/signaltools.py
+++ b/scipy/signal/signaltools.py
@@ -3021,6 +3021,8 @@ def resample(x, num, t=None, axis=0, window=None, domain='time'):
 
     x = np.asarray(x)
     Nx = x.shape[axis]
+    if np.all(np.imag(x) == 0.0):
+        x = np.real(x)
 
     # Check if we can use faster real FFT
     real_input = np.isrealobj(x)
```

For `num = 2`, the copying doesn't happen for complex inputs, may be that's why the output is different. I will dig a bit deeper to find out the root cause of this issue.

https://github.com/scipy/scipy/blob/e4b3e6eb372b8c1d875f2adf607630a31e2a609c/scipy/signal/signaltools.py#L3079

**Comment by mreineck** [comment]
I don't think there is a need for a quick fix like that ... probably better to find the root of the problem, as you say.
I'm only using the comparison of real vs. complex resampling in the test case because that was the shortest demonstrator.

**Comment by czgdp1807** [comment]
I observed that when we keep, Nyquist frequency for `N = min(num, X.shape[axis])` then inconsistencies go away. However there is one test in `test_signaltools.py` which fails. See the code below,
```python
x1 = np.array([1. + 0.j, 0. + 0.j])
y1_test = scipy.signal.resample(x1, 4) # y1_test = [0.5+0.j 0.5+0.j 0.5+0.j 0.5+0.j]
y1_true = np.array([1.+0.j, 0.5+0.j, 0.+0.j, 0.5+0.j])
print(x1, y1_test, y1_true)
```

Is there any way to find if `y1_test` is also a correct output for `x1` as input? @mreineck

**Comment by mreineck** [comment]
I think the proposed change is not the correct one, since the current result of the test in `test_signaltools.py` looks good, and should stay unchanged after fixing the original problem. Note that the problem only occurs for downsampling; all upsampling operations I tested were OK.

**Comment by mreineck** [comment]
If I had to guess, I suspect the problem lies somewhere around here:

https://github.com/scipy/scipy/blob/ba5801702ba8f5ef24ecdd5c4451d9bd7c2e661c/scipy/signal/signaltools.py#L3077-L3081

**Comment by czgdp1807** [comment]
May be. The problem is either with Nyquist frequency for N = 2 or the one you have said. Most probably the issue is due to incorrect handling of corner case, N = 2.

**Comment by DietBru** [comment]
You are correct @czgdp1807. The bug is in [these lines], which copy the negative frequency parts:
```python
sl[axis] = slice(-N//2, -N//2 + 1)
Y[tuple(sl)] += X[tuple(sl)]
```
When `N = min(num, Nx)` is `2` then `sl[axis] = slice(-1, 0, None)`. This does not copy
the value `X[-1]`, which it should. A corrected version would be:
```python
l1 = -N//2 + 1 if -N//2 + 1 < 0 else None 
sl[axis] = slice(-N//2, l1)
Y[tuple(sl)] += X[tuple(sl)]
```
It produces the correct `slice(-1, None, None)` for `N = 2`.

[these lines]: https://github.com/scipy/scipy/blob/d96ce7ab40f4b723be95f775437d89f7b1e4166a/scipy/signal/_signaltools.py#L3682

(I have a PR in the works which still needs a little more time to ripen ...)
